demos/http/http_demo_s3_upload/http_demo_s3_upload.c

```diff
@@ -0,0 +1,561 @@
+#define IOT_CORE_ALPN_PROTOCOL_NAME_LENGTH        ( sizeof( IOT_CORE_ALPN_PROTOCOL_NAME ) - 1 )
+
+/**
+ * @brief Field name of the HTTP Range header to read from server response.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD           "Content-Range"
+
+/**
+ * @brief Length of the HTTP Range header field.
+ */
+#define HTTP_CONTENT_RANGE_HEADER_FIELD_LENGTH    ( sizeof( HTTP_CONTENT_RANGE_HEADER_FIELD ) - 1 )
+
+/**
+ * @brief Delay in seconds between each iteration of the demo.
+ */
+#define DEMO_LOOP_DELAY_SECONDS                   ( 5U )
+
+/**
+ * @brief The length of the data in bytes to upload.
+ */
+#define DEMO_HTTP_UPLOAD_DATA_LENGTH              ( sizeof( DEMO_HTTP_UPLOAD_DATA ) - 1 )
+
+/**
+ * @brief The length of the HTTP GET method.
+ */
+#define HTTP_METHOD_GET_LENGTH                    ( sizeof( HTTP_METHOD_GET ) - 1 )
+
+/**
+ * @brief The length of the HTTP PUT method.
+ */
+#define HTTP_METHOD_PUT_LENGTH                    ( sizeof( HTTP_METHOD_PUT ) - 1 )
+
+/**
+ * @brief A buffer used in the demo for storing HTTP request headers and HTTP
+ * response headers and body.
+ *
+ * @note This demo shows how the same buffer can be re-used for storing the HTTP
+ * response after the HTTP request is sent out. However, the user can also
+ * decide to use separate buffers for storing the HTTP request and response.
+ */
+static uint8_t userBuffer[ USER_BUFFER_LENGTH ];
+
+/**
+ * @brief Represents header data that will be sent in an HTTP request.
+ */
+static HTTPRequestHeaders_t requestHeaders;
+
+/**
+ * @brief Configurations of the initial request headers that are passed to
+ * #HTTPClient_InitializeRequestHeaders.
+ */
+static HTTPRequestInfo_t requestInfo;
+
+/**
+ * @brief Represents a response returned from an HTTP server.
+ */
+static HTTPResponse_t response;
+
+/**
+ * @brief The host address string extracted from the pre-signed URL.
+ *
+ * @note S3_PRESIGNED_PUT_URL_LENGTH is set as the array length here as the
+ * length of the host name string cannot exceed this value.
+ */
+static char serverHost[ S3_PRESIGNED_PUT_URL_LENGTH ];
+
+/**
+ * @brief The length of the host address found in the pre-signed URL.
+ */
+static size_t serverHostLength;
+
+/**
+ * @brief The location of the path within the pre-signed URL.
+ */
+static const char * pPath;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Connect to HTTP server with reconnection retries.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created
+ * network context.
+ *
+ * @return EXIT_FAILURE on failure; EXIT_SUCCESS on successful connection.
+ */
+static int32_t connectToServer( NetworkContext_t * pNetworkContext );
+
+/**
+ * @brief Retrieve and verify the size of the S3 object that is specified in
+ * pPath.
+ *
+ * @param[in] pTransportInterface The transport interface for making network
+ * calls.
+ * @param[in] pPath The Request-URI to the objects of interest. This string
+ * should be null-terminated.
```

Comment:
```suggestion
 * must be null-terminated.
```
Because you are using strlen